File: 백준/Gold/4803. 트리/트리.cc
#include <iostream>
#include <vector>
using namespace std;

vector<int> v[501];
bool visited[501];

bool dfs(int x, int post_x) {
	visited[x] = true;

	for (int i = 0; i < v[x].size(); i++) {
		int next = v[x][i];
		if (next == post_x) continue;
		if (visited[next]) return false;
		if (!dfs(next, x)) return false;
	}
	return true;
}

int main() {
	cin.tie(NULL);
	ios::sync_with_stdio(false);

	int n, m;
	int caseidx = 0;

	while (1) {

		cin >> n >> m;
		if (n == 0 && m == 0) return 0;
		
		caseidx++;
		for (int i = 1; i < 501; i++) {
			v[i].clear();
			visited[i] = false;
		}

		int a, b;
		for (int i = 0; i < m; i++) {
			cin >> a >> b;
			v[a].push_back(b);
			v[b].push_back(a);
		}

		int cnt = 0;
		for (int i = 1; i <= n; i++) {
			if (!visited[i])
				if (dfs(i, 0)) cnt++;
		}

		if (cnt == 0) cout << "Case " << caseidx << ": No trees.\n";
		else if (cnt == 1) cout << "Case " << caseidx << ": There is one tree.\n";
		else if (cnt >= 2) cout << "Case " << caseidx << ": A forest of " << cnt << " trees.\n";
	}

	
	
}

File: 백준/Gold/4803. 트리/트리.java
import java.io.*;
import java.util.*;

public class Main {
	static int[] parents;
	static boolean[] isTree;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		StringBuilder sb = new StringBuilder();
		int t = 0;
		while (true) {
			t++;
			st = new StringTokenizer(br.readLine());

			int n = Integer.parseInt(st.nextToken());
			int m = Integer.parseInt(st.nextToken());

			if(n == 0 && m == 0) break;
			
			parents = new int[n + 1];
			isTree = new boolean[n + 1];
			for (int i = 1; i <= n; i++) {
				parents[i] = i;
				isTree[i] = true;
			}
			
			for (int i = 0; i < m; i++) {
				st = new StringTokenizer(br.readLine());
				int a = Integer.parseInt(st.nextToken());
				int b = Integer.parseInt(st.nextToken());
				
				union(a, b);
			}

			HashSet<Integer> set = new HashSet<>();
			for(int i = 1; i<=n; i++) {
				if(isTree[find(i)]) {
					set.add(find(i));
				}
			}
			
			int cnt = set.size();
			
			sb.append("Case ").append(t).append(": ");
			if(cnt > 1) {
				sb.append("A forest of ").append(cnt).append(" trees.");
			}
			else if(cnt == 1) {
				sb.append("There is one tree.");
			}
			else {
				sb.append("No trees.");
			}
			sb.append('\n');
		}
		System.out.println(sb);
	}
	
	static int find(int x) {
		if(x == parents[x])
			return x;
		return parents[x] = find(parents[x]);
	}
	
	static void union(int a, int b) {
		int aRoot = find(a);
		int bRoot = find(b);
		
		if(aRoot == bRoot) {
			isTree[bRoot] = false;
			return;
		}
		
		parents[aRoot] = bRoot;
		isTree[bRoot] = isTree[aRoot] && isTree[bRoot];
	}
}
